In mysqlxx::Pool: fix for reconnection problem

Akazz opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix


Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed problem when ClickHouse fails to resume connection to MySQL servers.


Detailed description / Documentation draft:
Under certain conditions CH may become unable to resume its connections to mysql when mysql server is used as a data source (external dictionaries, tables, etc.)

Akazz committed:
Remove stalled unrecoverable connection from Pool

Akazz committed:
For better compliance with mysql docs when calling mysql_real_connect()

Akazz committed:
A testing program for mysqlxx::Pool

alexey-milovidov commented on base/mysqlxx/Pool.cpp:
```diff
@@ -150,14 +151,22 @@ Pool::Entry Pool::tryGet()
+                return res;
+
+            /// This one is disconnected for good and so must be removed
+            ::delete *connection_it;  /// TODO: Manual memory management is awkward (matches allocConnection() method)
```
So, the fix is in these two lines?

Akazz replied:
Basically yes, it is important to get rid of such unrecoverable connections.

Akazz committed:
Tiny typo - major problem

Akazz committed:
Correct style to fix build

Akazz commented:
Failing [**Functional stateless tests**](https://clickhouse-test-reports.s3.yandex.net/17681/62649d0dd0777f3080319830601e4abc56f4d363/functional_stateless_tests_(memory).html#fail1) in 62649d0 seem to be flapping in *01304_direct_io*.

Generally, this PR looks safe to be merged in.

Akazz committed:
Empty commit

Akazz committed:
Added more connection logging to make tracking down problems simpler

Akazz committed:
Fixed clang-11 build

clang-11 generates style/readability warning
[readability-qualified-auto,-warnings-as-errors]

Akazz commented:
Tests are OK
